### PR TITLE
ACS-5487 - Track Total Hits on ES

### PR DESCRIPTION
**Pull request by evasques**
- SearchParameters - added trackTotalHits (int) attribute
 - SearchRequest - Added trackTotalHitsLimit (int) to the Limits attribute and mapped it to the SearchParameters
 - Changed the SearchRequest model in TAS to include a new RestRequestLimitsModel that has the new trackTotalHitsLimit attribute
 - SearchMapperTests to test the changes in the SearchParameters

### CODE REVIEW

File: data-model/src/main/java/org/alfresco/service/cmr/search/SearchParameters.java [hunk 1]
```diff
@@ -201,6 +201,8 @@ public enum Operator
 
     private String timezone;
     
+    private int trackTotalHits = 0;
```
Comment (tpage-alfresco) [inline data-model/src/main/java/org/alfresco/service/cmr/search/SearchParameters.java:204]: Please could we have a javadoc comment somewhere explaining what this field holds?

File: remote-api/src/test/java/org/alfresco/rest/api/search/SearchMapperTests.java [hunk 2]
```diff
@@ -755,23 +755,28 @@ public void fromLimits() throws Exception
...
+        searchMapper.fromLimits(searchParameters, new Limits(1000, null, null));
         assertEquals(LimitBy.NUMBER_OF_PERMISSION_EVALUATIONS, searchParameters.getLimitBy());
         assertEquals(1000, searchParameters.getMaxPermissionCheckTimeMillis());
         assertEquals(-1, searchParameters.getLimit());
         assertEquals(-1, searchParameters.getMaxPermissionChecks());
+
+        searchParameters = new SearchParameters();
+        searchMapper.setDefaults(searchParameters);
+        searchMapper.fromLimits(searchParameters, new Limits(null, null, 10));
+        assertEquals(10, searchParameters.getTrackTotalHits());
```
Comment (tpage-alfresco) [inline remote-api/src/test/java/org/alfresco/rest/api/search/SearchMapperTests.java:779]: This test method could be split into about five test methods.

File: data-model/src/main/java/org/alfresco/service/cmr/search/SearchParameters.java (hunk 1 above)
Comment (tpage-alfresco) [reply data-model/src/main/java/org/alfresco/service/cmr/search/SearchParameters.java:204]: From PMD:
> SearchParameters.java:204:	RedundantFieldInitializer:	Avoid using redundant field initializer for 'trackTotalHits'

File: packaging/tests/tas-restapi/src/main/java/org/alfresco/rest/search/RestRequestLimitsModel.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+package org.alfresco.rest.search;
+
+import org.alfresco.rest.core.IRestModel;
+import org.alfresco.utility.model.TestModel;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class RestRequestLimitsModel extends TestModel implements IRestModel<RestRequestLimitsModel>
+{
+    @JsonProperty(value = "entry")
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/main/java/org/alfresco/rest/search/RestRequestLimitsModel.java:35]: From PMD:
> RestRequestLimitsModel.java:35:	UnnecessaryAnnotationValueElement:	Avoid the use of value in annotations when its the only element

File: packaging/tests/tas-restapi/src/main/java/org/alfresco/rest/search/RestRequestLimitsModel.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    public RestRequestLimitsModel(Integer permissionEvaluationTime, Integer permissionEvaluationCount,
+            Integer trackTotalHitsLimit)
+    {
+        super();
+        this.permissionEvaluationTime = permissionEvaluationTime;
+        this.permissionEvaluationCount = permissionEvaluationCount;
+        this.trackTotalHitsLimit = trackTotalHitsLimit;
+    }
+
+    private Integer permissionEvaluationTime;
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/main/java/org/alfresco/rest/search/RestRequestLimitsModel.java:53]: From PMD:
> RestRequestLimitsModel.java:53:	FieldDeclarationsShouldBeAtStartOfClass:	Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.

File: packaging/tests/tas-restapi/src/main/java/org/alfresco/rest/search/RestRequestLimitsModel.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    private Integer permissionEvaluationCount;
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/main/java/org/alfresco/rest/search/RestRequestLimitsModel.java:54]: From PMD:
> RestRequestLimitsModel.java:54:	FieldDeclarationsShouldBeAtStartOfClass:	Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.

File: packaging/tests/tas-restapi/src/main/java/org/alfresco/rest/search/RestRequestLimitsModel.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    private Integer trackTotalHitsLimit;
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/main/java/org/alfresco/rest/search/RestRequestLimitsModel.java:55]: From PMD:
> RestRequestLimitsModel.java:55:	FieldDeclarationsShouldBeAtStartOfClass:	Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.

File: remote-api/src/test/java/org/alfresco/rest/api/search/SearchMapperTests.java (hunk 2 above)
Comment (tpage-alfresco) [reply remote-api/src/test/java/org/alfresco/rest/api/search/SearchMapperTests.java:779]: From PMD:
> SearchMapperTests.java:779:	JUnitAssertionsShouldIncludeMessage:	JUnit assertions should include a message

**Review by tpage-alfresco — COMMENTED** [review]
I added the new PMD issues as comments too - mainly so that you can see them more easily and assess whether the ruleset needs updating, or whether you would prefer to update the code.

File: data-model/src/main/java/org/alfresco/service/cmr/search/SearchParameters.java [hunk 7]
```diff
@@ -1641,6 +1647,19 @@ public void setIncludeMetadata(boolean includeMetadata)
...
     }
-    
+
+    public int getTrackTotalHits()
+    {
+        return trackTotalHits;
+    }
+
+    /**
+     * Configure the limit to track the total hits on search results
```
Comment (tpage-alfresco) [inline data-model/src/main/java/org/alfresco/service/cmr/search/SearchParameters.java:1657]: I still didn't get what the field is for - sorry. My guess is this:
```suggestion
     * Set a maximum value for the report of total hits. The reported number of hits will never exceed this limit even if more are found.
```

**Review by tpage-alfresco — APPROVED** [review]
I don't know that I understand the comment, but this is a definite improvement :+1:

**Review by jottley — APPROVED** [review]

**Pull request merged**